src/transformStateWithClones.js

```diff
@@ -7,7 +7,54 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const currentState = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(currentState, stateCopy, action.extraData);
+        break;
+      case 'removeProperties':
+        removeProperties(currentState, stateCopy, action.keysToRemove);
+        break;
+      case 'clear':
+        clear(currentState, stateCopy);
+        break;
+      default:
+        break;
+    }
+  }
+
+  return currentState;
+
+  function addProperties(newState, oldState, properties) {
+    Object.assign(oldState, properties);
+
+    const newStateObject = Object.assign({}, oldState);
+
+    newState.push(newStateObject);
+  }
```

Comment:
The `addProperties` function should create a new copy of the state object inside the function to avoid mutating the `oldState` argument. Instead of modifying `oldState`, create a new object that combines `oldState` and `properties`, then push this new state object to `newState`.